scr/src/main/java/org/apache/felix/scr/impl/logger/LogManager.java

```diff
@@ -0,0 +1,203 @@
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import org.osgi.framework.Bundle;
+import org.osgi.framework.BundleContext;
+import org.osgi.framework.BundleEvent;
+import org.osgi.framework.BundleListener;
+import org.osgi.framework.ServiceReference;
+import org.osgi.service.log.Logger;
+import org.osgi.service.log.LoggerFactory;
+import org.osgi.util.tracker.ServiceTracker;
+
+/*
+ * A class that tracks the LoggerFactory of a bundle and its associated Logger objects. 
+ * When a bundle is stopped and/or the service is unregistered, the logger objects are cleaned.
+ * 
+ * The basic technique is to use a facade. Instead of returning a log object, we return a facade. The 
+ * methods delegate to the actual logger object. If there is no logger object, we create one.
+ * 
+ * The LogDomain represents every bundle. Per LogDomain, we keep the facades. If the factory goes,
+ * we reset the facades.
+ */
+class LogManager extends ServiceTracker<LoggerFactory, LoggerFactory> {
+
+	// Guard is itself. Don't replace it with a ConcurrentHashMap since the
+	// LogDomain value is not stateless.
+
+	final Map<Bundle, LogDomain>	domains	= new HashMap<>();
+	final BundleContext				scrContext;
+	final AtomicBoolean				closed	= new AtomicBoolean(false);
+	volatile int					trackingCount;
+	volatile LoggerFactory			factory;
+
+	LogManager(BundleContext context) {
+		super(context, LoggerFactory.class, null);
+		this.scrContext = context;
+	}
+
+	private LogDomain getLogDomain(Bundle bundle) {
+		synchronized (domains) {
+			LogDomain domain = domains.get(bundle);
+			if (domain == null) {
+				domain = new LogDomain(bundle);
+				domains.put(bundle, domain);
+			}
+			return domain;
+		}
+	}
+
+	@Override
+	public void removedService(ServiceReference<LoggerFactory> reference, LoggerFactory service) {
+		if (!closed.get()) {
+			reset();
+		}
+	}
+
+	private void reset() {
+		for (LogDomain domain : domains.values()) {
+			domain.reset();
+		}
+	}
+
+	<T> T getLogger(Bundle bundle, String name, Class<T> type) {
+		return type.cast(getLogDomain(bundle).getLogger(name));
+	}
+
+	LoggerFactory getLoggerFactory() {
+		int trackingCount = getTrackingCount();
+		if (this.trackingCount < trackingCount) {
+			this.trackingCount = trackingCount;
+			factory = getService();
+		}
+		return factory;
+	}
+
+	/*
+	 * Tracks a bundle's LoggerFactory service
+	 */
+	class LogDomain
+			implements Closeable, BundleListener {
+
+		private final Bundle			bundle;
+		private final Set<LoggerFacade>	facades	= new HashSet<>();
+
+		LogDomain(Bundle bundle) {
+			this.bundle = bundle;
+			scrContext.addBundleListener(this);
+			open();
+		}
+
+		void reset() {
+			for (LoggerFacade facade : facades) {
+				facade.reset();
+			}
+		}
+
+		@SuppressWarnings("resource")
+		@Override
+		public void bundleChanged(BundleEvent event) {
+			if (event.getBundle() == bundle && event.getType() == BundleEvent.STOPPED && !closed.get()) {
```

Comment:
We can do this with a single `BundleListener`.  I suggest `LogManager` implement `BundleListener` because this method is managing the map from `LogManager`